### ISSUE TITLE
CN <--> hostname match

### ISSUE THREAD

**Issue by drwetter** [drwetter/testssl.sh#94]

**Comment by drwetter** [comment]
A PoC is in https://github.com/drwetter/testssl.sh/blob/master/utils/checkcert.sh but fails sometimes and needs more work

**Comment by dcooper16** [comment]
Hi @drwetter ,

I'm trying to do something for #276. Doing the additional test (TLSv1.1 w/o SNI) is easy, but the check of whether the same host is being hit is a bit more difficult. I want to say that if the hosts name ($NODE) appears in the certificate's CN or SAN, then then new certificate should be displayed, even if the CN and/or SAN aren't the same as in certificates that were previously seen. So, I looked at the existing code in certificate_info(), but I can't understand what it is trying to do:

```
if echo -n "$cn" | grep -q '^*.' ; then
     out " (wildcard certificate"
     cnfinding+="(wildcard certificate "
     if [[ "$cn" == "*.$(echo -n "$cn" | sed 's/^\*.//')" ]]; then
          out " match)"
          cnfinding+=" match)"
          wildcard=true
     else
          cnfinding+=" NO match)"
          cnok="INFO"
          #FIXME: we need to test also the SANs as they can contain a wild card (google.de .e.g) ==> 2.7dev
    fi
fi
```
What is the purpose of "if [[ "$cn" == "*.$(echo -n "$cn" | sed 's/^\*.//')" ]]; then"? The result of this is to either display "wildcard certificate match" or "wildcard certificate NO match," but the $cn is just being compared to a modified version of itself, so what does "match" or "NO match" mean? Was the first $cn supposed to be $NODE?

Here are the results I get for tools.ietf.org:

```
 TLS extensions (standard)    "renegotiation info/#65281" "EC point formats/#11" "session ticket/#35" "heartbeat/#15"
 Session Tickets RFC 5077     300 seconds (PFS requires session ticket keys to be rotated <= daily)
 SSL Session ID support       yes
 TLS clock skew               random values, no fingerprinting possible 
 Signature Algorithm          SHA256 with RSA
 Server key size              2048 bits
 Fingerprint / Serial         SHA1 B1114E17451F8626DA8C506DEA5A7EFFC9110EA8 / 9983BFB1CE0BE894
                              SHA256 4832567962EC34C2F22B8892CAB8F87065A9375B4A7F7B939D4E3D1B15467DD1
 Common Name (CN)             "*.tools.ietf.org" (wildcard certificate match) (works w/o SNI)
 subjectAltName (SAN)         "*.tools.ietf.org" "tools.ietf.org" 
 Issuer                       "Starfield Secure Certificate Authority - G2" ("Starfield Technologies, Inc." from "US")
 EV cert (experimental)       no 
 Certificate Expiration       187 >= 60 days (2015-10-13 12:42 --> 2016-11-30 18:34 -0500)
 # of certificates provided   4
 Chain of trust (experim.)    Ok   
 Certificate Revocation List  http://crl.starfieldtech.com/sfig2s1-17.crl
 OCSP URI                     http://ocsp.starfieldtech.com/
 OCSP stapling                --
```

Note that in this case the wildcard name in the CN does not match the server name. The server name appears in the SAN.

**Comment by drwetter** [comment]
That's piece of line just checks whether there's a wildcard match in the CN :-) The else statement probably won't  never be called. That code is legacy...

If one doesn't want to start from scratch I rather recommend checkcert.sh above. At least that code doesn't seem to fail, albeit it's not the greatest thing since sliced bread either. ;-)

**Comment by dcooper16** [comment]
Okay. I guess it's the word "match" that confuses me. Perhaps if the check passes, "(wildcard certificate)" should be added after the CN and if it doesn't, then nothing, rather than "(wildcard certificate match)" or "(wildcard certificate". (It seems that if the check fails, then "NO match)" is added to cnfinding, but it isn't printed.)

In any case, I went ahead and wrote some code to compare $NODE to the CN and SAN in $HOSTCERT. It uses the above check before trying a wildcard match against $NODE. So, I believe the fix for #276 in PR #371 is ready to go; there's just the question of whether my method of comparing two certificates is okay.

**Comment by drwetter** [comment]
> Okay. I guess it's the word "match" that confuses me. Perhaps if the check passes, "(wildcard 
> certificate)" should be added after the CN and if it doesn't, then nothing, rather than "(wildcard 
> certificate match)" or "(wildcard certificate". (It seems that if the check fails, then "NO match)" is 
> added to cnfinding, but it isn't printed.)

actually my remark with *match* was just intended as an indicator, not a real test. And that was only for wildcard.

I was thinking more on a line like
``Trust:         Ok via {SAN/CN}, SNI mandatory, otherwise <AAA> matches``

``Trust:         Ok via {SAN/CN wildcard}, works w/o SNI ``

I looked into #371 -- thx -- but not thoroughly enough yet. One thing I noted was ``sed 's/..DNS..`` might not be enough, see http://stackoverflow.com/questions/9185489/x509-subject-alternative-name-subjectaltname-ip-address-field . Actually I have a host with an IP address field in the SAN from a private CA, it helps not to throw warnings for ``https://<IPADDRESS>>``

Cheers, Dirk

**Comment by drwetter** [comment]
@dcooper16 : any chance to add a line like mentioned her https://github.com/drwetter/testssl.sh/issues/94#issuecomment-225301599 ?

**Comment by dcooper16** [comment]
I started to look into addressing this issue, and I discovered that the check for wildcard certificates in `compare_server_name_to_cert()` and `certificate_info()` is not complete. First, it would allow the wildcard character to match multiple labels (e.g., bar.foo.a.com would match *.a.com), whereas RFC 2818 says:
> Names may contain the wildcard character * which is considered to match any single domain name component or component fragment. E.g., \*.a.com matches foo.a.com but not bar.foo.a.com. f*.com matches foo.com but not bar.com.

Also, while it seems that browsers have decided to only accept wildcard names in certificates if the wildcard character represents the entire first domain label (e.g., \*.example.com), [RFC 2818](https://datatracker.ietf.org/doc/rfc2818/?include_text=1) (Section 3.1) and [RFC 6125](https://datatracker.ietf.org/doc/rfc6125/?include_text=1) (Section 6.4.3) explicitly allow the first label to contain other characters (e.g., b*z.example.net), and testssl.sh is not limited to browsers.

So, I would like to suggest changing the wildcard code to match [RFC 2818](https://datatracker.ietf.org/doc/rfc2818/?include_text=1) and [RFC 6125](https://datatracker.ietf.org/doc/rfc6125/?include_text=1). I have written a new function, `wildcard_match()`, that is intended to take a presented name (e.g., $NODE) and a name from a certificate (cn or SAN) and return 0 if both of the following are satisfied:

- The name from the certificate is a valid wildcard name (contains a "*" character in the first domain label); and
- The presented name is a wildcard match for the name from the certificate.

Below is a script that contains the `wildcard_match()` function along with some additional code to facilitate testing.

Please let me know if you'd like me to use this function as the basis for a PR that enhances the wildcard matching capabilities of testssl.sh based on [RFC 6125](https://datatracker.ietf.org/doc/rfc6125/?include_text=1).

```
#!/usr/bin/env bash

COLOR=${COLOR:-2}

# Return 0 if the name provided in arg2 is a wildcard name and it matches the name provided in arg1.
wildcard_match()
{
     # Comparisons of DNS names are case insenstive, so convert both names to uppercase.
     local servername=$1
     local certname=$2
     local temp part1 part2="" basename
     local -i i len_certname len_part1 len_basename
     local -i len_servername len_wildcard

     len_servername=${#servername}
     len_certname=${#certname}

     # Use rules from RFC 6125 to perform the match.

     # Assume the "*" in the wildcard needs to be replaced by one or more characters,
     # although RFC 6125 is not clear about that.
     [[ $len_servername -lt $len_certname ]] && return 1

     # If the first label in the DNS name begins "xn--", then assume it is an
     # A-label and don't try to perform a wildcard
     [[ "${certname:0:4}" == "xn--" ]] && return 1

     # Comparisons of DNS names are case insenstive, so convert both names to uppercase.
     certname="$(toupper "$certname")"

     # Get part of name preceding '*' or '.'
     basename="$(echo -n "$certname" | sed 's/^[A-Z0-9\-]*//g')"
     [[ "${basename:0:1}" != "*" ]] && return 1 # not a wildcard name
     len_basename=${#basename}
     i=$len_certname-$len_basename
     part1="${certname:0:i}"
     len_part1=${#part1}
     basename="${basename:1}"
     len_basename=$len_basename-1

     servername="$(toupper "$servername")"
     len_wildcard=$len_servername-$len_part1-$len_basename
     i=$len_servername-$len_basename

     # Check that initial part of $servername matches initial part of $certname
     # and that final part of $servername matches final part of $certname. 
     [[ "${servername:0:len_part1}" != "$part1" ]] && return 1
     [[ "${servername:i:$len_basename}" != "$basename" ]] && return 1

     # Check that part of $servername that matches "*" is all part of a single
     # domain label.
     [[ -n $(echo -n "${servername:$len_part1:$len_wildcard}" | sed 's/^[A-Z0-9\-]*//g') ]] && return 1

     return 0
}

toupper() {
     echo -n "$1" | tr 'a-z' 'A-Z'
}

out(){ 
          printf -- "%b" "${1//%/%%}"
}
outln() { out "$1\n"; }
pr_svrty_critical()  { [[ "$COLOR" -eq 2 ]] && out "\033[1;31m$1" || pr_bold "$1"; pr_off; }
pr_off()          { [[ "$COLOR" -ne 0 ]] && out "\033[m"; }
pr_bold()         { [[ "$COLOR" -ne 0 ]] && out "\033[1m$1" || out "$1"; pr_off; }

test_match()
{
    local servername=$1
    local certname=$2
    local -i ret

    wildcard_match "$servername" "$certname"
    ret=$?

    out "$servername "
    if [[ $ret -eq 0 ]]; then
         pr_bold "matches"
    else
         pr_svrty_critical "does not match" 
    fi
    out " $certname"
    [[ $ret -ne 0 ]] && out " (or $certname is not a wildcard name)"
    outln
}

test_match "fI9dod-56oTxRd8ds.exam.example.com" "fI9dod-56o*Rd8ds.exam.example.com"
test_match "fI9dod-56oTxRd8ds.exam.example.com" "fI9dod-56oTxRd8ds.exam.example.com"
test_match "fI9dod-56oTxRd8ds.exam.example.com" "*Rd8ds.exam.example.com"
test_match "fI9dod-56oTxRd8ds.exam.example.com" "*.exam.example.com"
test_match "fI9dod-56oTxRd8ds.exam.example.com" "fI9dod-56oTxRd8ds.*.example.com"
test_match "xn--56oTxRd8ds.exam.example.com" "xn--56*.exam.example.com"
test_match "foo.a.com" "*.a.com"
test_match "bar.foo.a.com" "*.a.com"
test_match "foo.com" "f*.com"
test_match "bar.com" "f*.com"
test_match "foo.example.com" "*.example.com"
test_match "bar.foo.example.com" "*.example.com"
test_match "example.com" "*.example.com"
test_match "baz1.example.net" "baz*.example.net"
test_match "foobaz.example.net" "*baz.example.net"
test_match "buzz.example.net" "b*z.example.net"
```

**Comment by drwetter** [comment]
David,

thx for putting so much thoughts into it.

Please go ahead!

Best, Dirk
-- 
Sent from my mobile. Excuse my brevity&typos

**Comment by drwetter** [comment]
Closed via df64e47
